### PR TITLE
Add DatabaseService heartbeat

### PR DESCRIPTION
**Pull request by marcoandredinis**
n the context of Teleport Discover, we must be able to known if there's any DatabaseService available to proxy a given Database resource.
If there's none available, we will offer a script for the user to run and install a DatabaseService which proxies the desired Database resource.

By DatabaseService, we mean the process that Teleport binary manages when the `teleport.yaml` config has the following section:
```yaml
db_service:
  enabled: "yes"
```

To accomplish this we are creating a new resource: DatabaseService.
The UI will fetch all DatabaseServices and check if there's any ResourceMatcher that matches the DatabaseLabels.

Previous PRs created the DatabaseService resource and its CRUD methods.

This PR adds an heartbeat for DatabaseServices similar to what we have for Databases.
There's also a new command to list DatabaseServices using `tctl`:
```
$ tctl get db_service --format text
Name                                 Resource Matchers                      
------------------------------------ -------------------------------------- 
a6065ee9-d5ee-4555-8d47-94a78625277b (Labels: <all databases>)              
d4e13f2b-0a55-4e0a-b363-bacfb1a11294 (Labels: env=[prod],aws-tag=[xyz abc]) 
```

Test using Teleport Cloud
```
dinis@lenix ~/p/cloudagents> tctl get db_services
kind: db_service
metadata:
  expires: "2022-12-21T18:05:10Z"
  id: 1671645310983808522
  name: 2a28d394-900c-42ea-a120-eed918e4526b
spec:
  resources:
  - labels:
      aws-tag:
      - xyz
      - abc
      env: prod
version: v1
dinis@lenix ~/p/cloudagents> tctl status
Cluster  marcoacme.cloud.gravitational.io                                        
Version  12.0.0-dev 
```

Part of #19032
Related #19363 #19469

### CODE REVIEW

**Review by ryanclark — APPROVED** [review]

File: tool/tctl/common/collection.go [hunk 1]
```diff
@@ -927,3 +927,52 @@ func (c *installerCollection) writeText(w io.Writer) error {
...
+	}
+	return r
+}
+
+func resourceMatchersToString(in []*types.DatabaseResourceMatcher) string {
+	resourceMatchersStrings := make([]string, len(in))
+
+	for i, resMatcher := range in {
+		labelsString := []string{}
+		for key, values := range *resMatcher.Labels {
```
Comment (greedy52) [inline tool/tctl/common/collection.go:947]: nitpick: check for nil

File: lib/srv/heartbeat.go [hunk 2]
```diff
@@ -133,6 +135,8 @@ const (
 	HeartbeatModeApp
 	// HeartbeatModeDB sets heatbeat to db
 	HeartbeatModeDB
+	// HeartbeatModeDatabaseService sets heatbeat mode to Databaseservice.
```
Comment (greedy52) [inline lib/srv/heartbeat.go:138]: nitpick:
```suggestion
	// HeartbeatModeDatabaseService sets heatbeat mode to database service.
```
or `DatabaseService`?

**Review by greedy52 — APPROVED** [review]

File: tool/tctl/common/resource_command.go [hunk 3]
```diff
@@ -1380,10 +1381,57 @@ func (rc *ResourceCommand) getCollection(ctx context.Context, client auth.Client
...
+		if len(dbServices) == 0 && rc.ref.Name != "" {
+			return nil, trace.NotFound("Database Service %q not found", rc.ref.Name)
+		}
+
+		return &databaseServiceCollection{databaseServices: dbServices}, nil
 	}
 	return nil, trace.BadParameter("getting %q is not supported", rc.ref.String())
 }
 
+func listDatabaseServices(ctx context.Context, client auth.ClientI, name string) ([]types.DatabaseService, error) {
```
Comment (smallinsky) [inline tool/tctl/common/resource_command.go:1399]: nit: can we implement this method in the `api/client`  like `(c *Client) GetDatabaseServices`  and reused it here ?

File: lib/srv/heartbeat.go (hunk 2 above)
Comment (marcoandredinis) [reply lib/srv/heartbeat.go:138]: `DatabaseService`
I wrote `DatabaseService` everywhere (I think) to ease the grepping for this particular resource type

But can refactor to `database service` if you think it's better

Comment (greedy52) [reply lib/srv/heartbeat.go:138]: Either way is good. As long as it's readable.

File: lib/srv/db/server_test.go [hunk 4]
```diff
@@ -239,7 +239,8 @@ func TestHeartbeatEvents(t *testing.T) {
 
 			require.NotNil(t, server)
 			require.Eventually(t, func() bool {
-				return atomic.LoadInt64(&heartbeatEvents) == test.heartbeatCount
+				// DatabaseService will emit an heartbeat, so we must increase the expected hearbeats by 1
+				return atomic.LoadInt64(&heartbeatEvents) == test.heartbeatCount+1
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply lib/srv/db/server_test.go]: Thank you!

I removed that section and adapted the test accordingly.
It seems to be working from a couple of tests I ran locally :crossed_fingers:

**Review by smallinsky — APPROVED** [review]

File: tool/tctl/common/resource_command.go (hunk 3 above)
Comment (marcoandredinis) [reply tool/tctl/common/resource_command.go]: Actually, we should be using ` apiclient.GetResourcesWithFilters` instead
It already iterates over all pages and return the entire list.

**Review by greedy52 — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

**Pull request merged**